### ISSUE TITLE
appSettings gives random values (old or new) depending on computer's performance

### ISSUE THREAD

**Issue by NordlingArt** [NativeScript/NativeScript#2061]
I am closing my old issue (#1925) about this because the problem I'm experiencing is more relevant to `appSettings` as a whole, and not just its `clear()` function.

**Test platform:** iOS

#### Description of the issue
`appSettings` will give me either an old value that's been set before, or the correct one. It's all random and it seems to depend on the performance of the computer. Under heavy load, I am experiencing this way too often and it is preventing me from further development of my app.

I have this in my settings-view-model:
```
var observable = require("data/observable").Observable;
var appSettings = require("application-settings");

var USER_EMAIL = "user-email";
var USER_PASSWORD = "user-password";

var settings = new observable();

Object.defineProperty(settings, "user_email", {
   get: function() { return appSettings.getString(USER_EMAIL, ""); },
   set: function(value) { appSettings.setString(USER_EMAIL, value); },
   enumerable: true,
   configurable: true
});
Object.defineProperty(settings, "user_password", {
   get: function() { return appSettings.getString(USER_PASSWORD, ""); },
   set: function(value) { appSettings.setString(USER_PASSWORD, value); },
   enumerable: true,
   configurable: true
});

module.exports = settings;
```

I have a login screen at startup which will check whether email and password is filled in or not, as in already set in the `appSettings`:
```
var settingsVM = require("../../shared/view-models/settings-view-model");
var observable = require("data/observable").Observable;
var frameModule = require('ui/frame');

var page;
var pageData = new observable({
  email: settingsVM.get("user_email"),
  password: settingsVM.get("user_password")
});

exports.loaded = function() {
  page = args.object;
  page.bindingContext = pageData;
  
  console.log("Email: " + settingsVM.get("user_email");
  console.log("Password": settingsVM.get("user_password");
  
  setTimeout(function() {
    console.log("Email: " + settingsVM.get("user_email");
    console.log("Password": settingsVM.get("user_password");
  }, 5000);
  
  if(!settingsVM.get("user_email") && !settingsVM.get("user_password")) {
    frameModule.topmost().navigate({
      moduleName: 'views/login/login',
      transition: { name: "fade" },
      clearHistory: true
    });
  } else {
    frameModule.topmost().navigate({
      moduleName: 'views/feed/feed',
      transition: { name: "fade" },
      clearHistory: true
    });
  }
};

exports.login = function() {
  settingsVM.set("user_email", pageData.email);
  settingsVM.set("user_password", pageData.password);
};
```

As you can see, I have console logs that check the `appSettings` values at page load, and a timeout delay of 5 seconds. While working on the app I have to restart it from time to time (of course), and for some reason first console logs will show empty values, or an OLD value that's been set way before. The timeout logs will then show either the actual values or just a random one once again (depending on computer load). Check out this scenario:

1. email and password is empty in `appSettings`.
2. I login and it changes to, say `test@test.com` and `test`.
3. I restart the app.
4. email and password is empty according to `appSettings`. Timeout logs show the real values.
5. I log in again with other credentials, `test2@test2.com` and `test2`.
6. I restart the app again.
7. email and password shows `test@test.com` and `test`. Timeout logs show `test2@test2.com` and `test2`.

I hope this explains the problem a bit better than what I explained in the old issue. What is causing this trouble and how can this be solved? [I found this](http://stackoverflow.com/questions/23443950/nsuserdefaults-values-are-lost-periodically) on StackOverflow which I think MIGHT be the same issue that I have.

Thanks in advance

**UPDATE:** From what I have gathered around the web regarding iOS's `NSUserDefaults` that `appSettings` uses, it does not seem to be a reliable feature to store user/app settings with ([stated here](https://forums.developer.apple.com/thread/15685)).

**Comment by NickIliev** [comment]
Hello @NordlingArt 

I have noticed that in your view-model.js you have the following code:

`get: function() { return appSettings.getString(USER_EMAIL, ""); },`

` get: function() { return appSettings.getString(USER_PASSWORD, ""); },`

getString should receive only single argument like: 

`get: function() { return appSettings.getString(USER_EMAIL); },`

` get: function() { return appSettings.getString(USER_PASSWORD); },`


However this is not causing your issue - I have tested your scenario as far as I can reproduce it and i was not able to reproduce your app behavior. Please keep in mind that if you test your device on simulator it will reset your app-settings every time you restart your app (for example with **tns run ios**) but if you test on real device you should be able to successfully login and redirect to feed-page.

As an better example of my testing scenario I have created this repository in which the whole case is reproduced.Here is the direct link [https://github.com/NickIliev/sample-LoginPage](https://github.com/NickIliev/sample-LoginPage)

**Comment by NordlingArt** [comment]
Hi @NickIliev According to the [documentation](https://docs.nativescript.org/ApiReference/application-settings/README.html), `getString` has a second optional argument that serves as a default value in case the string has no existing value.

You are right, the issues does not appear on a real device. After a lot of googling I realized that NSUserDefaults resets upon app rebuild. Seems like a logical reason for it to act up. There was also an issue with values still being there after clearing `appSettings`. That was fixed with a "Reset Content and Settings..." in the simulator.
